File: 邏輯迴歸/邏輯迴歸.py
#匯入套件
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

#讀取
data = pd.read_csv('social_network_ads.csv') 
x = data.iloc[:, [2,3]].values
y = data.iloc[:, 4].values

#訓練測式集
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y,
                                                test_size = 0.25, random_state = 0)
#y只能是0,1，不能夠特徵縮放
#對x做特徵縮放
from sklearn.preprocessing import StandardScaler
sc_x = StandardScaler()
x_train = sc_x.fit_transform(x_train)
x_test = sc_x.fit_transform(x_test)

#模型
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0, solver = 'liblinear')
classifier.fit(x_train, y_train)

#預測結果
y_pred = classifier.predict(x_test)

#評估性能(比較預設跟實際的差別)
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

#畫出訓練集結果
from matplotlib.colors import ListedColormap
x_set, y_set = x_train, y_train
#meshgrid是用兩個座標軸上的點在平面上畫網格(也可以是n維)
#min() - 1與max() + 1目的是使圖的邊緣留白(多一個單位)
#二維(年齡、平均薪水)
x1, x2 = np.meshgrid(np.arange(start = x_set[:, 0].min() - 1,
                    stop = x_set[:, 0].max() + 1, step = 0.01),
                    np.arange(start = x_set[:, 1].min() - 1,
                    stop = x_set[:, 1].max() + 1, step = 0.01))
#畫等高線圖
#填充顏色(紅色及綠色)
plt.contourf(x1, x2, classifier.predict(np.array([x1.ravel(), x2.ravel()]).T).reshape(x1.shape),
             alpha = 0.75, cmap = ListedColormap(('red', 'green')))
#設定值範圍
plt.xlim(x1.min(), x1.max())
plt.ylim(x2.min(), x2.max())
#迴圈(畫點圖)
#enumerate會產生兩個數值(第一個為序列數值，第二個維設定的參數)
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(x_set[y_set == j, 0], x_set[y_set == j, 1],
                c = ListedColormap(('orange','blue'))(i), label = j)
plt.title('classifier (train_set)')
plt.xlabel('Age')
plt.ylabel('Estimated Salary')
plt.legend()
plt.show()

#測試集
from matplotlib.colors import ListedColormap
x_set, y_set = x_test, y_test
x1, x2 = np.meshgrid(np.arange(start = x_set[:, 0].min() - 1,
                    stop = x_set[:, 0].max() + 1, step = 0.01),
                    np.arange(start = x_set[:, 1].min() - 1,
                    stop = x_set[:, 1].max() + 1, step = 0.01))
plt.contourf(x1, x2, classifier.predict(np.array([x1.ravel(), x2.ravel()]).T).reshape(x1.shape),
             alpha = 0.75, cmap = ListedColormap(('red', 'green')))
plt.xlim(x1.min(), x1.max())
plt.ylim(x2.min(), x2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(x_set[y_set == j, 0], x_set[y_set == j, 1],
                c = ListedColormap(('orange','blue'))(i), label = j)
plt.title('classifier (test_set)')
plt.xlabel('Age')
plt.ylabel('Estimated Salary')
plt.legend()
plt.show()

File: 邏輯迴歸/邏輯迴歸.R
#讀取資料
data = read.csv('social_network_ads.csv')
data = data[, 3:5]

#訓練測試集
library(caTools)
set.seed(123)
split = sample.split(data$Purchased, SplitRatio = 0.75)
train_set = subset(data, split == TRUE)
test_set = subset(data, split == FALSE)

#特徵縮放(對1, 2欄)
train_set[, 1:2] = scale(train_set[, 1:2])
test_set[, 1:2] = scale(test_set[, 1:2])

#建立模型
attach(data)
model = glm(Purchased ~ .,data = train_set, family = binomial)
summary(model)

#預測
prod_pred = predict(model, type = 'response', test_set[-3])
y_pred = ifelse(prod_pred > 0.5, 1, 0)

#製作confusion matrix
cm = table(test_set[, 3], y_pred)

#畫圖
library(ElemStatLearn)
#設置set方便以後要畫測試及時只需改一個值
set = train_set
#間隔越小，系統會跑越久
#-1及+1是使兩個邊界留白(多留一個單位)
x1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.05)
x2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.05)
#找出所有可能性列表(網格)
grid_set = expand.grid(x1, x2)
colnames(grid_set) = c('Age', 'EstimatedSalary')
prob_set = predict(model, type = 'response', newdata = grid_set)
y_grid = ifelse(prob_set > 0.5, 1, 0)
#畫出所有集的值
plot(set[, -3],
     main = 'Classifier (Training set)',
     xlab = 'Age', ylab = 'Estimated Salary',
     xlim = range(x1), ylim = range(x2))
#畫出分界
contour(x1, x2, matrix(as.numeric(y_grid), length(x1), length(x2)), add = TRUE)
#將網格區域畫出
points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'green', 'red'))
#將點上色
points(set, pch = 21, bg = ifelse(set[, 3] == 1, 'orange', 'blue'))

#測試集
set = test_set
x1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.05)
x2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.05)
grid_set = expand.grid(x1, x2)
colnames(grid_set) = c('Age', 'EstimatedSalary')
prob_set = predict(model, type = 'response', newdata = grid_set)
y_grid = ifelse(prob_set > 0.5, 1, 0)
plot(set[, -3],
     main = 'Classifier (test set)',
     xlab = 'Age', ylab = 'Estimated Salary',
     xlim = range(x1), ylim = range(x2))
contour(x1, x2, matrix(as.numeric(y_grid), length(x1), length(x2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'green', 'red'))
points(set, pch = 21, bg = ifelse(set[, 3] == 1, 'orange', 'blue'))
